### ISSUE TITLE
cURL error 60: SSL certificate problem: unable to get local issuer certificate

### ISSUE THREAD

**Issue by ondrejhlavacek** [keboola/http-extractor#31]
@yustme říká, že certifikát to má správnej

https://connection.eu-central-1.keboola.com/admin/utils/logs?file=2019/03/19/19/2019-03-19-19-57-38-5c913ba2a3994-exception
https://connection.eu-central-1.keboola.com/admin/utils/logs?file=2019/03/19/19/2019-03-19-19-57-38-5c913ba2ae51e-log

**Comment by ondrejhlavacek** [comment]
prej je to z https://opendata.sukl.cz/

**Comment by yustme** [comment]
je to https://opendata.sukl.cz/ . Jako certifikát využívají DigiCert. Pokud to skusím stáhnout pomocí wget, tak mi to hlásí stejnou chybu. Takže problém je s chybějícím certifikátem v systému. Možná by bylo fajn tam mít zaškrtáváko "skip ssl verification" v http extractoru. Jinak lidi nestáhnou nic, kde není platný SSL certifikát.

**Comment by odinuv** [comment]
lepsi by bylo tam ten certifikat pridat. To, ze nemas platny certifikat neco rika o duveryhodnosti tech dat

**Comment by tomasfejfar** [comment]
A vyřešilo by to něco jako https://packagist.org/packages/composer/ca-bundle ?

**Comment by tomasfejfar** [comment]
^Zkusím jestli to nevyřeší a kdyžtak můžeme vymýšlet dál.

**Comment by odinuv** [comment]
imho vyresi https://github.com/composer/ca-bundle/blob/master/res/cacert.pem#L2243

**Comment by tomasfejfar** [comment]
Hmm, tak ne :( https://travis-ci.org/keboola/http-extractor/builds/508849620?utm_source=github_status&utm_medium=notification

**Comment by tomasfejfar** [comment]
Ani v cli: `curl https://opendata.sukl.cz/soubory/SOD20190228/DLP20190228.zip --cacert /code/vendor/composer/ca-bundle/res/cacert.pem`

**Comment by tomasfejfar** [comment]
https://www.ssllabs.com/ssltest/analyze.html?d=opendata.sukl.cz

**Comment by tomasfejfar** [comment]
Takže se obávám, že bude potřeba tam ten "skip ssl verification" přidat. Musíme ho pojmenovat dostatečně složitě `very_unsafe_disable_ssl_check_i_know_what_i_am_doing`, aby to lidi nepoužívali automaticky a aby byla práce ho zapnout :) (v UI třeba opsaním textu vyrenderovaného v canvasu 😈 )

**Comment by odinuv** [comment]
mno to je zase zajeb

**Comment by ondrejhlavacek** [comment]
nebo to nedat do UI, pouze v power user režimu

**Comment by odinuv** [comment]
certifikat je v poradku, ale sukl ma spatne nakonfigurovany server - tady je to popsane vcetne odkazu na RFC:
https://serverfault.com/questions/774211/do-web-servers-send-the-certificate-chain-to-the-web-client
https://unix.stackexchange.com/questions/490608/how-to-make-gnu-linux-trust-a-certificate-thats-trusted-by-windows-out-of-the-b

projevuje se to takhle:
`openssl s_client -connect www.keboola.com:443 -prexit`:

```
root@f95ce47d9b31:/code# openssl s_client -connect www.keboola.com:443 -prexit
CONNECTED(00000003)
depth=2 C = BE, O = GlobalSign nv-sa, OU = Root CA, CN = GlobalSign Root CA
verify return:1
depth=1 C = BE, O = GlobalSign nv-sa, CN = AlphaSSL CA - SHA256 - G2
verify return:1
depth=0 OU = Domain Control Validated, CN = *.netlify.com
verify return:1
---
Certificate chain
 0 s:/OU=Domain Control Validated/CN=*.netlify.com
   i:/C=BE/O=GlobalSign nv-sa/CN=AlphaSSL CA - SHA256 - G2
 1 s:/C=BE/O=GlobalSign nv-sa/CN=AlphaSSL CA - SHA256 - G2
   i:/C=BE/O=GlobalSign nv-sa/OU=Root CA/CN=GlobalSign Root CA
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGWzCCBUOgAwIBAgIMdidMYjVGqvU1UiGeMA0GCSqGSIb3DQEBCwUAMEwxCzAJ
BgNVBAYTAkJFMRkwFwYDVQQKExBHbG9iYWxTaWduIG52LXNhMSIwIAYDVQQDExlB
...
```

a pro sukl `openssl s_client -connect opendata.sukl.cz:443 -prexit` je to 
```
root@f95ce47d9b31:/code# openssl s_client -connect opendata.sukl.cz:443 -prexit
CONNECTED(00000003)
depth=0 C = CZ, L = Praha, O = Statni ustav pro kontrolu leciv, CN = *.sukl.cz
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 C = CZ, L = Praha, O = Statni ustav pro kontrolu leciv, CN = *.sukl.cz
verify error:num=21:unable to verify the first certificate
verify return:1
---
Certificate chain
 0 s:/C=CZ/L=Praha/O=Statni ustav pro kontrolu leciv/CN=*.sukl.cz
   i:/C=US/O=DigiCert Inc/OU=www.digicert.com/CN=Thawte TLS RSA CA G1
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGYDCCBUigAwIBAgIQArXEfpG5sWU6E8tvYN/+jjANBgkqhkiG9w0BAQsFADBe
MQswCQYDVQQGEwJVUzEVMBMGA1UEChMMRGlnaUNlcnQgSW5jMRkwFwYDVQQLExB3
...
```

coz je spatne. Takze muzem zkusit napsat do SUKLu, at si opravi HTTPS - cc @yustme jestli s nima neco resi. Flag bych na to nepridaval - je uplne jedno jakej ma nazev a co musi uzivatel udelat, protoze jednou to nekdo zaklikne a uz to v te konfiguraci bude navzdy spatne. Kdyz uz tak by to mohla byt nova komponenta UNSAFE HTTP Extractor

**Comment by ErikZigo** [comment]
Custom komponentu bych to nedelal.
Nedavno jsme se u MySQL rozhodli to udelat primo do komponenty. V ui je to myslim skryty.
https://github.com/keboola/db-extractor-mysql/pull/96

**Comment by yustme** [comment]
Já bych tam přidal tu možnost nevalidního certifikátu. Kotel věcí má self-signed certifikát a ty jako mají smůlu ?

**Comment by tomasfejfar** [comment]
Já jsem taky spíš pro ten switch na "unsafe". Často to není o tom, jestli ty data jsou skutečně od toho, od koho čekáš, ale spíš o tom, aby nebyly odposlechnutelná po cestě. A na to self-signed certifikát imho úplně stačí.

**Comment by odinuv** [comment]
> A na to self-signed certifikát imho úplně stačí.

samozrejme, ze nestaci

**Comment by tomasfejfar** [comment]
Protože... ? Moje chápání je, že se self-signed certifikátem stále platí, že útočník nemůže přečíst cestou přenášená data (tj. např. nemusím se bát stahnout sensitive data na public wifině). Sice může udělat MITM útok a poslat jiná data, ale nemůže cestou rozšifrovat, to co posílá legitimní server. Což mi přijde jako ten good-enough scénář.

**Comment by odinuv** [comment]
ano, muze udelat MITM, a timpadem muze data rozsifrovat, jinak by to nebylo MITM. Udelat to muze diky tomu, ze si vytvorit uplne stejny self-signed certifikat jako legitimni server a server nijak neoveruje komu ty data poslal, takze end-user nic nepozna. 

lepsi argument proto to povolit, je, ze ten extractor stejne umoznuje stahnout data z http
(nicmene https bez kontroly je porad horsi nez http)

**Comment by yustme** [comment]
@odinuv pořád nevidím důvod, když zaškrnu checkbox "unsafe" a někdo mi podvrhne certifikát, tak se můžou stáhnout špatný data. Proto se ten checkbox jmenuje "unsafe" a ne "Don't worry it will be fine".

A nevím kolik útoků MITM se vám, kdy povedlo udělat. Ale MITM s http je mnohem jednodušší než s https.

Stále jsem pro 
- [ ] skip ssl validation (unsafe)
 + tomu můžeme přidat vyskakovací warning, že to někdo může podvrhnout.

Jinak vyhodíme ze hry všechny self-signed certifikáty, které jsou podle mě lepší než čisté HTTP

**Comment by odinuv** [comment]
ano, vsak jsem psal:

> lepsi argument proto to povolit, je, ze ten extractor stejne umoznuje stahnout data z http

> Jinak vyhodíme ze hry všechny self-signed certifikáty, které jsou podle mě lepší než čisté HTTP

nejsou, a ta tvoje veta sama vysvetluje proc nejsou. Protoze jsou uplne stejne spatny jako HTTP, ale spousta lidi si mysli, ze jsou lepsi a maji falesnej pocit bezpeci.

**Comment by tomasfejfar** [comment]
https://security.stackexchange.com/a/99835/25116

**Comment by odinuv** [comment]
btw, mezitim mi ze suklu napsali, ze to opravili

**Comment by yustme** [comment]
> https://security.stackexchange.com/a/99835/25116

Mít jako argument odpověd na SO, která má `-2` není dobrý tah :smile:

**Comment by odinuv** [comment]
no kazdopadne kdyz to ted funguje, tak tohle muzeme dat on-hold a vyprdnout se na to dokud se to zase neobjevi, ne?
